File: 680. Valid Palindrome II/Python/Valid Palindrome II.py
class Solution:
    def validPalindrome(self, s):
        """
        :type s: str
        :rtype: bool
        """

        def is_palindrome(s, left, right):
            while left < right:
                if s[left] != s[right]:
                    return False
                left += 1
                right -= 1
            return True

        left = 0
        right = len(s) - 1
        while left < right:
            if s[left] != s[right]:
                return is_palindrome(s, left + 1, right) or is_palindrome(s, left, right - 1)
            left += 1
            right -= 1
        return True

File: 680. Valid Palindrome II/C++/Valid Palindrome II.cpp
class Solution {
public:
    bool validPalindrome(string s) {
        class Utils {
        public:
            bool isPalindrome(string &s, int left, int right) {
                while (left < right) {
                    if (s[left] != s[right]) {
                        return false;
                    }
                    left++;
                    right--;
                }
                return true;
            }
        };
        int left = 0;
        int right = s.length() - 1;
        Utils utils;
        while (left < right) {
            if (s[left] != s[right]) {
                return utils.isPalindrome(s, left + 1, right) || utils.isPalindrome(s, left, right - 1);
            }
            left++;
            right--;
        }
        return true;
    }
};